osu.Game/Rulesets/Objects/SliderPathExtensions.cs

```diff
@@ -25,6 +29,71 @@ public static void SnapTo<THitObject>(this THitObject hitObject, IDistanceSnapPr
         /// <param name="sliderPath">The <see cref="SliderPath"/>.</param>
         /// <param name="positionalOffset">The positional offset of the resulting path. It should be added to the start position of this path.</param>
         public static void Reverse(this SliderPath sliderPath, out Vector2 positionalOffset)
+        {
+            var controlPoints = sliderPath.ControlPoints;
+            var originalControlPointTypes = controlPoints.Select(p => p.Type).ToArray();
+
+            controlPoints[0].Type ??= PathType.Linear;
+
+            // Inherited points after a linear point should be treated as linear points.
+            controlPoints.Where(p => sliderPath.PointsInSegment(p)[0].Type == PathType.Linear).ForEach(p => p.Type = PathType.Linear);
+
+            double[] segmentEnds = sliderPath.GetSegmentEnds().ToArray();
+            double[] distinctSegmentEnds = segmentEnds.Distinct().ToArray();
+
+            // Remove control points at the end which do not affect the visual slider path ("invisible" control points).
+            if (segmentEnds.Length >= 2 && Math.Abs(segmentEnds[^1] - segmentEnds[^2]) < 1e-10 && distinctSegmentEnds.Length > 1)
```

Comment:
Okay so first of all, if there is a way to produce a zero-length slider in the lazer editor, then that is a bug. We have been tightening that down so that it shouldn't be possible to create one.

Secondly the distinct thing has me baffled because it looks wrong and brittle. First of all, distinct is prone to failures like "the 2nd and the 5th control point out of 8 have the same position". Or like "this control point is 0.00001 to the left of the other". And finally, the code seems to be implicitly relying on the fact that `.Distinct()` doesn't reorder items, which is [not guaranteed](https://learn.microsoft.com/en-us/dotnet/api/system.linq.enumerable.distinct?view=net-7.0):

Reply:
> Okay so first of all, if there is a way to produce a zero-length slider in the lazer editor, then that is a bug. We have been tightening that down so that it shouldn't be possible to create one.

I submitted a bug report (#24559).
> So if the goal here is to check that the last N control points are sorta in the same place [...]

Just to make sure there is no misunderstanding: The goal of this line https://github.com/ppy/osu/blob/cd706734636cd3d8da5bef4e9078c4b4d70cd5f9/osu.Game/Rulesets/Objects/SliderPathExtensions.cs#L41 was to get the `segmentEnds` without the duplicates *at the end of the array*, not to *actually* remove control points which are in the same place as others. Since `GetSegmentEnds()` gives us the same number for all control points beyond the sliderend, this can be used to detect how many control points the reversed slider shall have.

 That being said, this
> "the 2nd and the 5th control point out of 8 have the same position"

and this
> the code seems to be implicitly relying on the fact that .Distinct() doesn't reorder items, which is [not guaranteed](https://learn.microsoft.com/en-us/dotnet/api/system.linq.enumerable.distinct?view=net-7.0)

is still valid, so I replaced `.Distinct()` with a custom method (`truncateEndingDuplicates()`) instead.